test/controllers/tasks_controller_test.rb

```diff
@@ -88,37 +88,121 @@
   describe "edit" do
     it "can get the edit page for an existing task" do
-      skip
+      #skip
       # Your code here
+        get edit_task_path(task.id)
+        must_respond_with :success
     end
     
     it "will respond with redirect when attempting to edit a nonexistant task" do
-      skip
+      #skip
       # Your code here
+      get edit_task_path(-1)
+      must_respond_with :redirect
     end
   end
   
   # Uncomment and complete these tests for Wave 3
   describe "update" do
     # Note:  If there was a way to fail to save the changes to a task, that would be a great
     #        thing to test.
-    it "can update an existing task" do
       # Your code here
+    before do
+        Task.create(name: "Testing", description: " Testing description", completed_at: "Testing completes task")
+    end
+    let (:new_task_hash) {
+        {
+            task: {
+                name: "Make beds",
+                description: "Make all 3 beds",
+                completed_at: "Everyday",
+            },
+        }
+      }
+    it "can update an existing task" do
+      id = Task.first.id
+      expect {
+        patch task_path(id), params: new_task_hash
+      }.wont_change "Task.count"
+
+      must_redirect_to tasks_path
+
+      task = Task.find_by(id: id)
+      expect(task.name).must_equal new_task_hash[:task][:name]
+      expect(task.description).must_equal new_task_hash[:task][:description]
+      expect(task.completed_at).must_equal new_task_hash[:task][:completed_at]
+
     end
     
     it "will redirect to the root page if given an invalid id" do
       # Your code here
+      id = -1
+
+      expect {
+        patch task_path(id), params: new_task_hash
+      }.wont_change "Task.count"
+
+      must_respond_with :redirect
+
     end
   end
   
   # Complete these tests for Wave 4
   describe "destroy" do
     # Your tests go here
-    
+    it "can destroy a task" do
+      # Arrange
+      delete_task = Task.new(name: "Wash car", description: "Clean the inside and outside", completed_at: "Sunday")
+
+      delete_task.save
+      id = delete_task.id
+
+      # Act
+      expect {
+        delete task_path(id)
+
+        # Assert
+      }.must_change "Task.count", -1
+
+      must_respond_with :redirect
+      must_redirect_to tasks_path
+    end
+
+    it "will respond with not_found for invalid ids" do
+      expect {
+        delete task_path(-1)
+      }.wont_change "Task.count"
+
+      must_respond_with :not_found
+    end
   end
   
   # Complete for Wave 4
   describe "toggle_complete" do
     # Your tests go here
+    it "marks a task as completed" do
+      completed_task = Task.new(name: "Wash car", description: "Clean the inside and outside", completed_at: "time at completion")
+
+      completed_task.save
+      id = completed_task.id
+
+    # Act
+      expect {
+      patch toggle_complete_path(task.id)
+
+      # Assert
+      }.must_change "Task.count"
```

Comment:
I am unsure why this test passes. Take a look. We should expect that marking complete does not change the number of tasks in the database.